src/datahike/migrate.clj

```diff
@@ -17,23 +23,127 @@
+                        (-> (apply d/datom item)
+                            ;; convert Instant to Date (previously experienced bug)
+                            (update :v instance-to-date)
+                            ;; convert Float to Double (previously reported bug)
+                            (update :v instance-to-double)))
+        add-datom (fn [item]
+                     ;; skip schema datoms
+                    (if filter-schema?
+                      (when (-> item (nth 1) (str "0000") (subs 0 4) (not= ":db/")) (.put q (prepare-datom item)))
+                      (.put q (prepare-datom item))))
+        drain-queue (fn []
+                      (let [acc (java.util.ArrayList.)]
+                          ;; max required otherwise if previous write is slow too many in transaction
+                        (.drainTo q acc 200000)
+                        (vec acc)))]
+      ;; prevent all the datoms that failed from being logged
+    (timbre/merge-config! {:min-level [[#{"datahike.writing"} :fatal] [#{"datahike.writer"} :fatal]]})
+
+    (async/thread
+      (process-cbor-file
+       path
+       add-datom
+       (fn []
+         (reset! processed true))))
+
+    (async/thread
+      (timbre/info "Starting import")
+      (loop []
+        (Thread/sleep 100) ; batch writes for improved performance
+        (when (or @processed (not sync?))
+          (let [datoms (drain-queue)]
+            (try
+              (timbre/debug "loading" (count datoms) "datoms")
+              (swap! txn-count + (count datoms))
+
+              ;; in sync mode the max-tx will be as the test expect
+              ;; in async mode the max-tx will be tx-max + 1
+              (when sync?
+                (swap! conn assoc :max-tx @tx-max))
+
+              (if load-entities?
+                (api/load-entities conn datoms) ;; load entities is faster for large datasets
+                (api/transact conn datoms)) ;; transact is slow but preserves max-tx id
+              (catch Exception e
+                  ;; we can't print the message as it contains all the datoms
+                (timbre/error "Error loading" (count datoms))))))
+        (when (and (>= @txn-count @datom-count) @processed)
+            ;; stop when we've transacted all datoms
+          (when-not sync?
+            (swap! conn assoc :max-tx @tx-max))
+          (timbre/info "\nImported" @datom-count "datoms in total. \nTime elapsed" (- (System/currentTimeMillis) star-time) "ms")
+          (reset! stop true))
+        (when (not @stop) (recur))))
+
+    (when sync?
+      (loop []
+        (Thread/sleep 100)
+        (timbre/info "remaining" (- @datom-count @txn-count))
+        (when (not @stop) (recur))))
+
+      ;; allow the user to stop or monitor the ingestion thread with this atom
+    (fn []
+      {:complete? @stop
+       :preprocessed? @processed
+       :total-datoms @datom-count
+       :remaining (- @datom-count @txn-count)})))
+
+(comment
```

Comment:
@alekcz Thank you so much for helping with this! Overall it looks good to me and it is great that this code performs well for you. However, I cannot evaluate this comment block, stub-server is not in scope and the alias d is already loaded. Could you turn this into a test?